src/Option.purs

```diff
@@ -1229,6 +1230,49 @@ set proxy value = modify proxy go
   go :: forall a. a -> value
   go _ = value
 
+-- | Changes a key with the given value (if it exists) to an option.
+-- | As with `set`, the key must already exist in the option.
+-- |
+-- | E.g.
+-- | ```PureScript
+-- | someOption :: Option.Option ( foo :: Boolean, bar :: Int )
+-- | someOption = Option.empty
+-- |
+-- | anotherOption :: Option.Option ( foo :: Boolean, bar :: Int )
+-- | anotherOption = Option.setMay (Data.Symbol.SProxy :: _ "bar") (Just 31 someOption
+-- | ```
+-- |
+-- | The `proxy` can be anything so long as its type variable has kind `Symbol`.
+-- |
+-- | It will commonly be `Data.Symbol.SProxy`, but doesn't have to be.
+setMay ::
+  forall label option option' proxy value.
+  Data.Symbol.IsSymbol label =>
+  Prim.Row.Cons label value option' option =>
+  proxy label ->
+  Data.Maybe.Maybe value ->
+  Option option ->
+  Option option
+setMay proxy vMay def = modify proxy go def
+  where
+  go :: value -> value
+  go optVal = case vMay of
+    Data.Maybe.Just v -> v
+    Data.Maybe.Nothing -> optVal
+
+-- | A convenience function calling `setMay` that can be used to iteratively
+-- | mutate an existing `Option`, where mutations may occur in any order.
+-- |
+maySetOptState :: forall label option option' proxy value.
+  Data.Symbol.IsSymbol label =>
+  Prim.Row.Cons label value option' option =>
+  proxy label ->
+  Data.Maybe.Maybe value ->
+  Option option ->
+  Control.Monad.State.State (Option option) Unit
```

Comment:
I'm down with this, but maybe we can change the type around a bit and use `Control.Monad.State.Class.MonadState (Option option) _`. That should allow it to unify with `Control.Monad.State.State (Option option) Unit` as well as `Control.Monad.State.StateT (Option option) _ Unit` and any other data type that implements the type class.

Reply:
Alternatively, we relax to `Control.Monad.State.StateT (Option option) _ Unit`.

Reply:
Although my exposure to Monad Transformers is minimal, I'm generally in favor of the more generic option, so `MonadState` sounds good here.

Reply:
As I'm working through the documentation of `maySetOptState`, I'm not really sure how it'd be used. Do you have a couple of examples of when you turn to it to solve a problem?

Reply:
I guess what I'd assume is that you wouldn't pass the `Option _` in directly each time:

```PureScript
someOption :: Option.Option ( foo :: Boolean, bar :: Int )
someOption = Option.empty

foo :: Control.Monad.State.State (Option.Option ( foo :: Boolean, bar :: Int )) Unit
foo = do
  Option.maySetOptState (Data.Symbol.SProxy :: _ "bar") (Data.Maybe.Just 31) someOption
```
but would use the `Option _` that's in the state:

```PureScript
foo :: Control.Monad.State.State (Option.Option ( foo :: Boolean, bar :: Int )) Unit
foo = do
  Option.maySetOptState (Data.Symbol.SProxy :: _ "bar") (Data.Maybe.Just 31)
```
I assume you wanted it the way it is for a reason, but I'm not sure what that reason is. I think knowing that would help write better documentation.

Reply:
You may have a valid point, as my experience with the state monad is fairly limited, but [here's an example of how I had been using it](https://github.com/labordynamicsinstitute/metajelo-ui/blob/0e93f825d46df54bb981cfad8526a2e972186f80/src/Metajelo/UI.purs#L129-L142), and there are similar examples in that same file. So I think what you guessed about my usage is correct.

Reply:
Jawesome! Thanks for the examples. There's some things I'm thinking:

* If we changed the type of `maySetOptState` to use the `Control.Monad.State.State _ _`, that specific example could be re-written to be:
    ```PureScript
    foo = do
      newRec <- pure $ execState (do
        Opt.maySetOptState (SProxy :: _ "identifier_opt") (Just idOpt)
        Opt.maySetOptState (SProxy :: _ "identifier") idMay
        Opt.maySetOptState (SProxy :: _ "date") dateMay
        Opt.maySetOptState (SProxy :: _ "lastModified") xsdDateMay


        Opt.maySetOptState (SProxy :: _ "_numRelIds") (Just _numRelIds)
        Opt.maySetOptState (SProxy :: _ "relId_opts") relIdOpts
        Opt.maySetOptState (SProxy :: _ "relatedIdentifiers") relIdsMay


        Opt.maySetOptState (SProxy :: _ "_numSupProds") (Just _numSupProds)
        Opt.maySetOptState (SProxy :: _ "supProd_opts") supProdOpts
        Opt.maySetOptState (SProxy :: _ "supplementaryProducts") supProdsMay
        ) recOpt
    ```
    That seems like it would make the specific example a bit easier to deal with.
* Given that we'd now have a way to set values in a `Control.Monad.State.State _ _`, we might want to provide parallels for the other things you'd do: getting, modifying, etc. Probably would move to a different module so the namespacing would take care of things a bit better.
* Attemping to see the intent of that section of code, it seems like using `Control.Monad.State.State _ _` is primarily because `Option.set` only works with a single field at a time. In fact, most of the `Option` module only works on a single field. It seems like if there was a way to set an entire record of values, that'd be useful. You might write that example as:
    ```PureScript
    foo = do
      newRec <- Opt.setMay
        { identifier_opt: Just idOpt
        , identifier: idMay
        , date: dateMay
        , lastModified: xsdDateMay


        , _numRelIds: Just _numRelIds
        , relId_opts: relIdOpts
        , relatedIdentifiers: relIdsMay


        , _numSupProds: Just _numSupProds
        , supProd_opts: supProdOpts
        , supplementaryProducts: supProdsMay
        }
        recOpt
    ```
    Not only does this mean you don't have to do anything with `Control.Monad.State.State _ _`, you also don't have to do anything with `Data.Symbol.SProxy _`. Since it's a record, it degenerates to allow setting a single field: `Option.setMay { bar: 13 } someOption`. What do you think about that?

Reply:
> is primarily because Option.set only works with a single field at a time.

Yes, that's right, and using State seemed preferable to shadowing (if possible) or creating new binding names for each `set`. 

I think the simplified `setMay` syntax, using a record, would be amazing!

FYI, One thing I'd eventually like to do is to also work on a way to derive some of the Option types I'm using in the code from the underlying records.

Reply:
> FYI, One thing I'd eventually like to do is to also work on a way to derive some of the Option types I'm using in the code from the underlying records.

Can you talk about this some more? I _think_ what you want might already be possible, but I don't want to jump to conclusions.

Reply:
The background for my current usage, though I'm sure it is more general, is to create the "view-model", where the "model" itself is defined as a Record (or the row-type underlying the record type, at least) . The view-model often has the current state of the GUI stored in it, which often needs to be managed explicitly by the app developer in certain frameworks (if not all of them). 

As an example, in that same file I linked earlier (UI.purs), we have:

```purescript
-- | ViewModel for MetajeloRecord
type MetajeloRecordExtra r = (
  identifier_opt :: Opt.Option (M.BaseIdRows ())
, _numRelIds :: Int
, relId_opts :: PartialRelIds
, _numSupProds :: Int
, supProd_opts :: PartialProds
| r
)
-- | Decorated state (Model + ViewModel) for MetajeloRecord
type MetajeloRecordRowOpts = MetajeloRecordExtra M.MetajeloRecordRows
```

I don't expect to be able to derive `MetajeloRecordExtra` automatically, as only the app developer can really know what state needs to be preserved for the view model in every case, or at the very best, it would be framework and context specific.


Another example is something that *might* be possible to improve on, though I'm far from certain:

```purescript
-- | ViewModel for SupplementaryProduct
type SupplementaryProductExtra r = (
  basicMetadata_opt :: Opt.Option M.BasicMetadataRows
, resourceID_opt :: Opt.Option (M.BaseIdRows ())
, resourceType_opt :: Opt.Option M.ResourceTypeRows
, _numFormats :: Int
, resMdsOpts_opt :: Opt.Option ResourceMetadataSourceRowOpts
, locationOpts_opt :: Opt.Option LocationRowOpts
| r
)
-- | Decorated state (Model + ViewModel) for SupplementaryProduct
type SupplementaryProductRowOpts =
  SupplementaryProductExtra M.SupplementaryProductRows
```

In this case, what we're doing is tacking on the view-model Options for sub-components that had been previously generated elsewhere. The good news is, these only have to be tacked on for the immediate level down in the record hierarchy. For instance, any other `Option` types defined in the `Opt.Option LocationRowOpts` are already included. 

It isn't too bad to deal with as it is, but I think if this could be mitigated further, it would further improve the use of Opt.Option as a generic, pure-FP way to handle view-models.

Reply:
I'm sorry, I didn't really understand those examples. Don't feel obligated to rephrase or explain more, but I'll try and understand what you're saying if you do.

Reply:
I'm really not sure it is feasible anyway, but I agree I should explain this in more detail. Was I clear on the part regarding how an Option is a nice way to create a [view-model](https://stackoverflow.com/questions/11064316/what-is-viewmodel-in-mvc) from a Record (the model)?